Assets/HoloToolkit/Input/Scripts/ControllerVisualizer.cs

```diff
@@ -1,58 +1,92 @@
     /// </summary>
     [RequireComponent(typeof(SetGlobalListener))]
-    public class ControllerVisualizer : MonoBehaviour, ISourceStateHandler, ISourceRotationHandler, ISourcePositionHandler
+    public class ControllerVisualizer : MonoBehaviour, ISourcePositionHandler, ISourceRotationHandler
     {
         [Tooltip("Use a model with the tip in the positive Z direction and the front face in the positive Y direction.")]
         [SerializeField]
-        private GameObject controllerModel;
+        protected GameObject leftControllerOverride;
+        [Tooltip("Use a model with the tip in the positive Z direction and the front face in the positive Y direction.")]
+        [SerializeField]
+        protected GameObject rightControllerOverride;
+        [Tooltip("Use this to override the indicator used to show the user's touch location on the touchpad. Default is a sphere.")]
+        [SerializeField]
+        protected GameObject touchpadTouchedOverride;
+
+        [Tooltip("This shader will be used on the loaded GLTF controller model. This does not affect the above overrides.")]
+        public Shader GLTFShader;
 
-        // This will be used to keep track of our controllers, indexed by its unique source ID.
-        private Dictionary<uint, GameObject> controllerDictionary;
+#if !UNITY_EDITOR && UNITY_WSA
+        // This is used to get the renderable controller model, since Unity does not expose this API.
+        private SpatialInteractionManager spatialInteractionManager;
+#endif
+
+        // This will be used to keep track of our controllers, indexed by their unique source ID.
+        private Dictionary<uint, ControllerInfo> controllerDictionary;
 
         private void Start()
         {
-            if (controllerModel == null)
+            controllerDictionary = new Dictionary<uint, ControllerInfo>();
+
+#if !UNITY_EDITOR && UNITY_WSA
+            // Since the SpatialInteractionManager exists in the current CoreWindow, this call needs to run on the UI thread.
+            UnityEngine.WSA.Application.InvokeOnUIThread(() =>
             {
-                Debug.Log("Please make sure to add a controller model in the Inspector for ControllerVisualizer.cs on " + name + ".");
-            }
+                spatialInteractionManager = SpatialInteractionManager.GetForCurrentView();
+                if (spatialInteractionManager != null)
+                {
+                    spatialInteractionManager.SourceDetected += ControllerVisualizer_SourceDetected;
+                    spatialInteractionManager.SourceLost += ControllerVisualizer_SourceLost;
 
-            controllerDictionary = new Dictionary<uint, GameObject>();
+                    spatialInteractionManager.SourceUpdated += ControllerVisualizer_SourceUpdated;
+                }
+            }, true);
+#else
+            // Since we're using non-Unity APIs, this will only run in a UWP app.
+            Debug.Log("ControllerVisualizer only works in UWP.");
```

Comment:
Can we not make this script work with overridden models in the editor? Obviously not the ones from Windows update.